src/main/java/pl/joegreen/sergeants/simulator/GameMap.java

```diff
@@ -0,0 +1,240 @@
+     */
+    private int[] getCitiesDiff() {
+        List<Integer> allCities = Arrays.stream(tiles)
+                .filter(tile -> tile.getClass() == City.class)
+                .map(Tile::getTileIndex)
+                .collect(Collectors.toList());
+
+        allCities.add(0, allCities.size()); //second element = wrong
+        allCities.add(0, 0);//first element = zero correct
+        return allCities.stream().mapToInt(i -> i).toArray();
+    }
+
+    /**
+     * Returns all generals, visible and fogged. This is not a problem since bot will still receive tile as fogged (-3).
+     *
+     * @return list of tileIndex where general is, array index is player index
+     */
+    private int[] getGenerals() {
+        return Arrays.stream(tiles)
+                .filter(tile1 -> tile1.getClass().equals(General.class))
+                .sorted((t1, t2) -> t1.getPlayerIndex() - t2.getPlayerIndex())
+                .mapToInt(Tile::getTileIndex)
+                .toArray();
+    }
+
+    private boolean isVisible(int playerIndex, Tile tile) {
+        boolean selfTile = tile.getPlayerIndex() == playerIndex;
+        return selfTile || getSurroundingTiles(tile).stream().anyMatch(t -> t.getPlayerIndex() == playerIndex);
+    }
+
+    private List<Tile> getSurroundingTiles(Tile tile) {
+        //this could be prettier
+        List<Tile> ret = new ArrayList<>();
+        getTileLeftOf(tile).ifPresent(leftTile -> {
+            ret.add(leftTile);
+            getTileAbove(leftTile).ifPresent(ret::add);
+            getTileBelow(leftTile).ifPresent(ret::add);
+        });
+        getTileRightOf(tile).ifPresent(t -> {
+            ret.add(t);
+            getTileAbove(t).ifPresent(ret::add);
+            getTileBelow(t).ifPresent(ret::add);
+        });
+        getTileAbove(tile).ifPresent(ret::add);
+        getTileBelow(tile).ifPresent(ret::add);
+        return ret;
+    }
+
+    private Optional<Tile> getTileAbove(Tile tile) {
+        int tileIndex = tile.getTileIndex() - width;
+        return tileIndex >= 0 ? Optional.of(tiles[tileIndex]) : Optional.empty();
+    }
+
+    private Optional<Tile> getTileBelow(Tile tile) {
+        int tileIndex = tile.getTileIndex() + width;
+        return tileIndex < tiles.length ? Optional.of(tiles[tileIndex]) : Optional.empty();
+    }
+
+    private Optional<Tile> getTileRightOf(Tile tile) {
+        boolean isMostRight = tile.getTileIndex() % width == width - 1;
+        return isMostRight ? Optional.empty() : Optional.of(tiles[tile.getTileIndex() + 1]);
+    }
+
+    private Optional<Tile> getTileLeftOf(Tile tile) {
+        int tileIndex = tile.getTileIndex();
+        boolean isMostLeft = tileIndex == 0 || tileIndex % width == 0;
+        return isMostLeft ? Optional.empty() : Optional.of(tiles[tileIndex - 1]);
+    }
+
+    public int getHalfTurnCounter() {
+        return halfTurnCounter;
+    }
+
+    public void saveGameAsJson() {
+        //pls don't look at this code
```

Comment:
I think we could skip merging "saving the replay to file" feature for now and also skip merging the html file that shows the replays. As you noticed yourself the code in the current state has some disadvantages. Skipping the code-quality fact that javascript code is printed from java strings it also always writes into the the same js file, so users can only always see the last replay and if they run new simulation the old replay will be lost. 

I believe that the best way to implement it would be to let user pass the file handle when starting a simulation and write the replay json there. Also I believe that instead of creating a new replay format it would be best to save as .gioreplay http://dev.generals.io/replays - I've seen on discord that people are creating some replay browsers based on that format so it would be probably the best way to go. It would also fit nicely if https://github.com/greenjoe/sergeants/issues/5  is implemented.